test/runtests.jl

```diff
@@ -246,3 +246,28 @@ end
     @test isa(tuplegroups.A.i, Tuple)
     @test isa(tuplegroups.B.j, Tuple)
 end
+
+struct FreeParam{T,P<:NamedTuple} <: AbstractParam{T}
+    parent::P
+end
+FreeParam(nt::NT) where {NT<:NamedTuple} = begin
+    ModelParameters._checkhasval(nt)
+    FreeParam{typeof(nt.val),NT}(nt)
+end
+FreeParam(val; kwargs...) = FreeParam((; val=val, kwargs...))
+FreeParam(; kwargs...) = FreeParam((; kwargs...))
+
+@testset "spefify param type" begin
```

Comment:
Since you seem to agree that the `FreeParam`/`FixedParam` is not a good use-case for this feature, I would like to see an example use-case where this feature actually is actually necessary in contrast to using table manipulations, i.e. something where simply adding an extra field to `Param` wouldn't cut it or would be prohibitively difficult. I can't really think of one off the top of my head, but maybe you can.